Add tag groups and in file metadata

Numerlor opened the pull request:
Introduces tag groups described in #1545.

When matching a tag with a group, the group is matched with no minimum amount of chars (unlike a tag that has to have 3 chars of its name) to allow a short invocation name to be used and possibly prevent some conflicts with tag names where multiple tags would be suggested instead of a s tag being sent.

Additionally tags now support metadata through a frontmatter block like in [meta files in branding](https://raw.githubusercontent.com/python-discord/branding/main/events/earth_day/meta.md)
Currently, the `restricted_to` key accepts a list of roles to restric the tag to, and the `embed` key is passed to `discord.Embed.from_dict`, this allows customization of the embed for individual tags

The module was redesigned a bit to separate the logic of all the fuzzy matching and name parsing from the cog itself.

Numerlor commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +243,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+        await LinePaginator.paginate(result_lines, ctx, embed=embed, max_lines=15, empty=False, footer_text=FOOTER_TEXT)
+
+    async def list_tags_in_group(self, ctx: Context, group: str) -> None:
+        """Send a paginator with all tags under `group`, that are accessible by `ctx.author`."""
```
As it is, listing the top level group alone isn't possible because it lacks a group to look it up with. Would some kind of a string (`None`, `nogroup` or something like that) be worth using here to list them alone instead of only being accessible through the whole list

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,214 @@
+class COOLDOWN(enum.Enum):
+    """Sentinel value to signal that a tag is on cooldown."""
+
+    obj = object()
```
What's the difference to this and `COOLDOWN = object()`?

I assume this has the benefit of `__repr__` right?

Numerlor replied:
I used an Enum because I couldn't find a way to type the sentinel alone

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,214 @@
+            self,
+            ctx: Context,
+            keywords: str,
+            matching_tags: list[tuple[TagIdentifier, Tag]],
```
We define this type a lot, isn't there a way to create an alias. As in:
```py
tag_pair = tuple[TagIdentifier, Tag]
```
Then this code can become:
```suggestion
            matching_tags: list[tag_pair],
```
May not be a huge benefit, but I thought I'd mention it.

ToxicKidz replied:
Although type aliases should be in PascalCase

Numerlor replied:
Yes, that'd be nicer, although I don't think `TagPair` is a fitting name as I'd expect a pair of `Tag` objs from that, not sure on a good name myself

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
 
-        Tags are on cooldowns on a per-tag, per-channel basis. If a tag is on cooldown, display
-        nothing and return True.
+        If the requested tag is on cooldown or no suggestions were found, return None.
```
This claims that if the tag is on cool-down that None will be returned.

Numerlor replied:
Reworded things a bit in e40cd9840c188210a7cfde6bffbf08ce563c2a11

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            if tag.accessible_by(ctx.author)
+        ]
+
+        if (tag := self._tags.get(tag_identifier)) is None:
```
I'd honestly place this outside of the if-statement to make it clear that `tag` is used after the if-statement, even though Python doesn't have strict scoping like JS for example.

Numerlor replied:
I think this was a leftover from when I was prototyping and had it in an elif, moved to a normal assignment in e40cd9840c188210a7cfde6bffbf08ce563c2a11

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            if not filtered_tags:
+                return None
+            suggested_tags_text = "\n".join(
+                f"**\N{RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK}** {identifier}"
```
I assume that this is the Unicode name for `»`?

Either replace this with that character, or change the usage seen a bit further up with this. For consistency.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+    async def list_all_tags(self, ctx: Context) -> None:
+        """Send a paginator with all loaded tags accessible by `ctx.author`, groups first, and alphabetically sorted."""
+        def tag_sort_key(tag_item: tuple[TagIdentifier, tag]) -> str:
+            ident = tag_item[0]
```
We may want to use tuple assignment here? Feel free to leave that as it is though.

Numerlor replied:
Do you mean doing `group, name = tag_item[0]` to avoid the else case?

Bluenix2 replied:
I was originally thinking about doing `ident, _ = tag_item` since `tag_item` should be treated more like a pair than a tuple.

But that sounds great doing `group, name = tag_item[0]`, I didn't think about TagIdentifier being a named tuple.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            return group+ident.name
+
+        result_lines = []
+        current_group = object()
```
Is this also a sentinel value? Can you not use a empty string (I know that None is used when a tag doesn't have a group).

Numerlor replied:
I suppose an empty string could be used, but I don't think there's any benefit of using it instead of a plain object like the current use

Bluenix2 replied:
Well there's a benefit in annotating the code

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+                    result_lines.append(f"\n\N{BULLET} **{identifier.group}**")
+                else:
+                    result_lines.append("\n\N{BULLET}")
+                current_group = identifier.group
```
You could move this like next to the comment about "A new group began".

Numerlor replied:
Moved the assignment to the comment and changed the uses to use the name directly instead of identifier.group in 5cbb7d20713c0a7027c2e0909ddcdb6e02dbddc2

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+
+    async def list_tags_in_group(self, ctx: Context, group: str) -> None:
+        """Send a paginator with all tags under `group`, that are accessible by `ctx.author`."""
+        embed = Embed(title=f"**Tags under *{group}***")
```
Can an embed title be bold?

Numerlor replied:
Looks like it can, but other embed titles were changed to not be bold and I think it was a bit harder to read so I removed the bolding in 96a8390290f379a90fa0f966b74993dd1ddf9f44

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            for identifier, tag in self._tags.items()
+            if identifier.group == group and tag.accessible_by(ctx.author)
+        )
+        await LinePaginator.paginate(tag_lines, ctx, embed, footer_text=FOOTER_TEXT, empty=False, max_lines=15)
```
On line 340 you use a kwarg for `embed`, shouldn't this also be a kwarg then or is that unnecessary?

Numerlor replied:
Changed the other use to not use a kwarg in e3b9924920a32e19adfa1274cfdbc4cf9ddb874a

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
-                    await ctx.send(embed=Embed.from_dict(tag['embed'])),
-                    [ctx.author.id],
-                )
+        Returns True if something was sent, or if the tag is on cooldown.
```
Could use `response`, or `message` as opposed to `something`.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            tag_name: TagNameConverter = None,
+    ) -> bool:
+        """
+        Get a specified tag, or a list of all tags if no tag is specified.
```
Should probably mention that this can display a tag group.

Numerlor replied:
Reworded things a bit in e40cd98

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+                )
+            return True
+        else:
+            return False
```
Could probably write this thing as:
```py
if embed is None:
    return False

if embed is not COOLDOWN.obj:
    await wait_for_deletion(...)

return True
```

Numerlor replied:
It looks a bit nicer, but I'm not sure if it's as clear on when True is returned

Bluenix2 replied:
Could fix that with some comments, like `If we reached here it was successful` or `Wait for the cooldown message to be deleted before returning True`.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +246,163 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+
+
+def extract_tag_identifier(string: str) -> TagIdentifier:
+    """Create a `TagIdentifier` instance from beginning of `string`."""
```
```suggestion
    """Create a `TagIdentifier` instance from the beginning of `string`."""
```

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,217 @@
+    group: Optional[str]
+    name: str
+
+    def get_fuzzy_score(self, fuzz_tag_identifier: TagIdentifier) -> float:
```
What's wrong with simply calling that `identifier` or `match`?

Numerlor replied:
Nothing wrong, but I wanted to make it clear that the method is fuzzy matching the passed in identifier with self as the target

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,217 @@
+        self.content = post.content
+        self.metadata = post.metadata
+        self._restricted_to: set[int] = set(self.metadata.get("restricted_to", ()))
+        self._cooldowns: dict[discord.TextChannel, float] = {}
```
Can't redis be used for this?

Numerlor replied:
What purpose would it serve? We gather the data from the tag file and after that it's unchanged at runtime

Bluenix2 replied:
Specifically I am talking about the cooldowns but that may not be big enough to experience a benefit?

Numerlor replied:
I don't think they're important enough to persist if that's what you mean

Bluenix2 replied:
No, Redis will not store them persistently.

Redis is an in-memory cache (key-value store). Pretty much exactly like this dict.

I would like to hear others thought about it, as I am no expert on Redis, but I suggest you look into it if you're unaware what Redis is.

Numerlor replied:
I'm aware of what it is, but I fail to see how it could be used here

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,217 @@
+                # Try fuzzy matching with only a name first
+                suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
+
+            if not len(tag_identifier.name) < 3:
```
```suggestion
            if len(tag_identifier.name) > 3:
```

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +249,168 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            self,
+            ctx: Context,
+            tag_identifier: TagIdentifier,
+    ) -> Optional[Union[Embed, Literal[COOLDOWN.obj]]]:
```
Pretty much the same thing, but we avoid some unnecessary nesting
```suggestion
    ) -> Union[Embed, Literal[COOLDOWN.obj], None]:
```

Numerlor replied:
I find the Optional a bit nicer as it's immediately obvious that it either returns None or some other value
I can change it if others agree though

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +249,168 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+
+        tag = self._tags.get(tag_identifier)
+        if tag is None:
+            if len(filtered_tags) == 1:
```
This can become an `and`.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +249,168 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+                tag_identifier = filtered_tags[0][0]
+                tag = filtered_tags[0][1]
+
+        if tag is not None:
```
Should we perhaps add a comment why this can't be an `else:`?

Numerlor replied:
I think it should be clear with `tag` being changed on the line right before the condition

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +249,168 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
-                    description="**There are no tags in the database!**",
-                    colour=Colour.red()
-                ))
+        if tag_name is None:
```
Make this an `elif` to be clear that either of these happen.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,125 +249,168 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
     bot.add_cog(Tags(bot))
+
+
+def extract_tag_identifier(string: str) -> TagIdentifier:
```
How does moving this to a classmethod sound?

Numerlor replied:
Sounds good, made it a constructor of the class in 359a17c00bb036f9a90fe7681592dee29f84c806

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,227 @@
+                group_score = 1
+            else:
+                # Ignore tags without groups if the identifier contains a group
+                return .0
```
Something I am a bit on the fence about, do we prefer `0.0` here?

wookie184 replied:
If I was writing it myself i'd probably write `0.0`, but both ways are valid and I don't think our style guide indicates any preferred way, so probably doesn't need changing.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,227 @@
+
+    def on_cooldown_in(self, channel: discord.TextChannel) -> bool:
+        """Check whether the tag is on cooldown in `channel`."""
+        return channel in self._cooldowns and self._cooldowns[channel] > time.time()
```
How does this work over midnight, considering it is only time.

wookie184 replied:
`time.time` is seconds since the epoch, https://docs.python.org/3/library/time.html#time.time

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,123 +259,156 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+    ) -> bool:
+        """
+        When arguments are passed in:
+        If a single argument is given and it matches a group name, list accessible all tags from that group.
```
I think this sentence needs revising.

Numerlor replied:
I'm open to any suggestions.
Is "If a single argument matching a group name is given, list accessible all tags from that group." better?

Bluenix2 replied:
Specifically I was first talking about "list accessible all tags", but I like what you suggested better than what is currently there. So "If a single argument matching a group name is given, list all accessible tags from that group." it is

Numerlor replied:
Oh I didn't even notice that, will fix

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,227 @@
+    current, index = 0, 0
+    _search = REGEX_NON_ALPHABET.sub("", search.lower())
+    _targets = iter(REGEX_NON_ALPHABET.split(target.lower()))
+    _target = next(_targets)
+    try:
+        while True:
+            while index < len(_target) and _search[current] == _target[index]:
+                current += 1
+                index += 1
+            index, _target = 0, next(_targets)
+    except (StopIteration, IndexError):
+        pass
```
```suggestion
    for _target in _targets:
        while index < len(_target) and _search[current] == _target[index]:
            current += 1
            index += 1
        index = 0
```

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,227 @@
+
+def _fuzzy_search(search: str, target: str) -> float:
+    """A simple scoring algorithm based on how many letters are found / total, with order in mind."""
+    current, index = 0, 0
```
Perhaps we can move this down and remove the index assignment, we only need it in the loop.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,123 +259,156 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
-                )
+    async def list_all_tags(self, ctx: Context) -> None:
+        """Send a paginator with all loaded tags accessible by `ctx.author`, groups first, and alphabetically sorted."""
+        def tag_sort_key(tag_item: tuple[TagIdentifier, tag]) -> str:
```
```suggestion
        def tag_sort_key(tag_item: tuple[TagIdentifier, Tag]) -> str:
```

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,225 @@
+        """Check whether `member` can access the tag."""
+        return bool(
+            not self._restricted_to
+            or self._restricted_to & {role.id for role in member.roles}
```
This is actually a bit sad when you think about it haha

Discord.py uses an internal array to store role IDs attached to a member, `.roles` is a property that gets all of the roles from cache by their ID. We then use a set comprehension to convert it to a set of.. role IDs.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,225 @@
+    _search = REGEX_NON_ALPHABET.sub("", search.lower())
+    _targets = iter(REGEX_NON_ALPHABET.split(target.lower()))
+
+    current, index = 0, 0
+    for _target in _targets:
+        while index < len(_target) and _search[current] == _target[index]:
+            current += 1
+            index += 1
+        index = 0
```
```suggestion
    current = 0
    for _target in _targets:
        index = 0
        while index < len(_target) and _search[current] == _target[index]:
            current += 1
            index += 1
```

Bluenix2 reviewed (commented):
Looks good otherwise

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,123 +261,156 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+
+            self.bot.stats.incr(
+                f"tags.usages"
+                f"{'.' + tag_identifier.group.replace('-', '_') if tag_identifier.group else ''}"
```
Oh I thought I deleted this comment, but yes it would behave differently. https://github.com/python-discord/bot/pull/1663#discussion_r697746775 would be equivalent.

But I am not that picky on the subject, only the usage of `f"tags.usages"` being an f-string though I guess that leads to consistency.

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,229 @@
+
+    def get_fuzzy_score(self, fuzz_tag_identifier: TagIdentifier) -> float:
+        """Get fuzzy score, using `fuzz_tag_identifier` as the identifier to fuzzy match with."""
+        if self.group is None:
+            if fuzz_tag_identifier.group is None:
+                # We're only fuzzy matching the name
+                group_score = 1
+            else:
+                # Ignore tags without groups if the identifier contains a group
+                return .0
+        else:
+            if fuzz_tag_identifier.group is None:
+                # Ignore tags with groups if the identifier does not have a group
+                return .0
+            else:
+                group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
```
I was able to simplify this code a bit considering there's 3 cases:
- One has a group and the other doesn't (XOR)
- Identical groups (potentially both None)
- Different strings and should be fuzzy searched

```suggestion
        if self.group is None != fuzzy_tag_identifier.group is None:
            # Ignore tags without groups if the identifier has a group and vice versa
            return .0

        if self.group == fuzzy_tag_identifier.group:
            # Completely identical, or both None
            group_score = 1
        else:
            group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
```

Numerlor replied:
The  `self.group is None != fuzzy_tag_identifier.group is None:` comparison won't work. we'd need to do something like (not sure if there's any nicer way to check that either of  them is None and they aren't equal)
```suggestion
        if (self.group is None or fuzz_tag_identifier.group is None) and self.group != fuzz_tag_identifier.group:
            # Ignore tags without groups if the identifier has a group and vice versa.
            return .0
        if self.group == fuzz_tag_identifier.group:
            # Completely identical, or both None.
            group_score = 1
        else:
            group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
```
I think that's a bit easier to read overall but the first condition could be confusing at a glance

Bluenix2 replied:
No `self.group is None != fuzzy_tag_identifier.group is None:` will work, because they become bools. Bools only have 2 states, True or False. So we have 4 cases:
- both `self.group` and `fuzzy_tag_identifier.group` are None (`True != True` *False*)
- `self.group` is None and `fuzzy_tag_identifer.group` is set (`True != False` *True*)
- `self.group` is set and `fuzzy_tag_identifier.group` is None (`False != True` *True*)
- both `self.group` and `fuzzy_tag_identifier.group` are set (`False != False` *False*)

We want to return when it is True, which is exaclty what is happening here.

Numerlor replied:
With how it's chained, it impossible for the condition to be True `self.group is None != fuzzy_tag_identifier.group is None` creates a `!=` comparison between `None` and `fuzzy_tag_identifier.group` which is then compared to be equal to `None`. The long equivalent (and how python handles it) would be 
`self_group is None and None != fuzzy_tag_identifier_group and fuzzy_tag_identifier_group is None`

```py
In [149]: for self_group, fuzzy_tag_identifier_group in itertools.product((None, "ba"), repeat=2):
     ...:     print(self_group, fuzzy_tag_identifier_group)
     ...:     print(self_group is None != fuzzy_tag_identifier_group is None)
     ...:     print((self_group is None or fuzzy_tag_identifier_group is None) and self_group != fuzzy_tag_identifier_g
     ...: roup)
     ...:     print()
     ...:
None None
False
False

None ba
False
True

ba None
False
True

ba ba
False
False
```

Bluenix2 replied:
Oh I see, so just add the parenthesis: `(self.group is None) != (fuzzy_tag_identifier.group is None)`

Numerlor replied:
Ah, yes that should work, sorry I couldn't make sense of the comparison before

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,229 @@
+
+    def on_cooldown_in(self, channel: discord.TextChannel) -> bool:
+        """Check whether the tag is on cooldown in `channel`."""
+        return channel in self._cooldowns and self._cooldowns[channel] > time.time()
```
This makes two lookups, what about the following instead?
```suggestion
        if cooldown := self._cooldowns.get(channel):
            return cooldown > time.time()

        return False
```

Numerlor replied:
I think doing the containment is a bit more readable and less verbose; I'm not too worried about the two checks as they're negligible in runtime to pretty much anything else that's happening in the bot.

We could also go with -inf as the default for a get and compare with that but again imo doing a normal containment check is the most explicit to show what's going on there

Bluenix2 replied:
I think that actually sounds like a great idea
```suggestion
        return self._cooldowns.get(channel, float('inf')) > time.time()
```

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,229 @@
         base_path = Path("bot", "resources", "tags")
+
         for file in base_path.glob("**/*"):
             if file.is_file():
```
Technically we can skip a level of indentation if we invert this

Numerlor replied:
The nesting is not too bad and there's nothing after it so I think it's fine as is

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -24,145 +29,229 @@
-    def _get_tags_via_content(self, check: Callable[[Iterable], bool], keywords: str, user: Member) -> list:
+    def get_fuzzy_matches(self, tag_identifier: TagIdentifier) -> list[tuple[TagIdentifier, Tag]]:
+        """Get tags with identifiers similar to `tag_identifier`."""
+        if tag_identifier.group is None:
+            if len(tag_identifier.name) < 3:
+                return []
+            else:
+                return self._get_suggestions(tag_identifier)
+        else:
+            if len(tag_identifier.group) < 3:
+                suggestions = []
+            else:
+                # Try fuzzy matching with only a name first
+                suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```
```suggestion
        if len(tag_identifier.name) < 3 or len(tag_identifier.group) < 3:
            return []
    
        if tag_identifier.group is None:
            return self._get_suggestions(tag_identifier)
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```

Numerlor replied:
these do not act the same, the current implementation allows groups shorter than 3 chars to be used to invoke tags as long as the name is longer than 3 and matches

Bluenix2 replied:
Hmm, I see now. What about this instead?

We're removing a level of nesting in the if-statements introducing several dimensions to keep track, by flattening them we get a clear case of when to do what
```suggestion
        if len(tag_identifier.name) < 3:
            return []
        elif tag_identifier.group is None:
            return self._get_suggestions(tag_identifier)

        if len(tag_identifier.group) < 3:
            # No point in fuzzy matching if the group is less than 3 characters
            suggestions = []
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```
Like this we have a pretty clear view..

- If the name is less then 3 characters we return immediately
- If the group is None we can just fuzzy match immediately and return
*I cut them in half here because the first two exit early, the second two continue down the function*
- If the group is less then 3 characters there's no point in fuzzy matching it and just continue with an empty list
- If name is longer then 3 characters, and we have a group over 3 character we fuzzy match the group and continue

Numerlor replied:
This won't let us match only with a name if a tag is matched with a long group but a short name, I'm on the edge of that being worth it for the increased code complexity.
On one hand it's relatively unlikely to happen as `!<name> <2chars>` would have to be invoked (maybe an emoji would be an actual case where that can happen) but then the code is also not *that* much less complex without the first if taking care of that case and then the name length check at the end

Bluenix2 replied:
> This won't let us match with only a name if a tag is matched with a long group but a short name

I see now, so we just remove that if-statement and leave it to the code below this. That way we will at least match for groups yeah.
```suggestion
        if tag_identifier.group is None:
            return self._get_suggestions(tag_identifier)

        if len(tag_identifier.group) < 3:
            # No point in fuzzy matching if the group is less than 3 characters
            suggestions = []
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```

Numerlor replied:
That would skip the name length check altogether, allowing fuzzying of invocations like `!de` which may just be a typo

Bluenix2 replied:
No it wouldn't, check the code below this comment.

Numerlor replied:
The code below won't run when the group is None

Bluenix2 replied:
Ah right, so we can special-case it and just add a second `and len(tag.name) >= 3:`

Numerlor replied:
With that `None` groups would go beyond the check for short names

Bluenix2 replied:
I am starting to see why the if-statements look exactly like they do- because I tried to write a suggestion for the code to be exactly the same. That said I came up with something like this:
```py
        if tag_identifier.group is None and len (tag_identifier.name) >= 3:
            return self._get_suggestions(tag_identifier)

        if tag_identifier.group is None or len(tag_identifier.group) < 3:
            # No point in fuzzy matching if the group is less than 3 characters or None
            suggestions = []
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```
But at this point, this suggestion can be refactored to consider the if-statements below it.

Bluenix2 replied:
```
        suggestions = []

        if tag_identifier.group is not None and len(tag_identifier.group) < 3:
            # Try fuzzy matching with only a name first
            suggestions += self._get_suggestions(TagIdentifier(None, tag_identifier.group))

            if len(tag_identifier.name) >= 3:
                suggestions += self._get_suggestions(tag_identifier)

            return suggestions
```
I can't use the GitHub suggestions for this, because that would duplicate code. But what do you think about this?

Numerlor replied:
That should work according to my tests after fixing the inequality in the first if, pushed it in ba2a74ac18f9e40fad917d51ffa2238a340416c0

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,123 +261,155 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+        Otherwise display the tag if one was found for the given arguments, or try to display suggestions for that name
 
-                self.bot.stats.incr(f"tags.usages.{tag['title'].replace('-', '_')}")
+        With no arguments, list all accessible tags
```
```suggestion
        With no arguments, list all accessible tags.
```

Bluenix2 commented on bot/exts/info/tags.py:
```diff
@@ -172,123 +261,155 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+    ) -> bool:
+        """
+        If a single argument matching a group name is given, list all accessible tags from that group
+        Otherwise display the tag if one was found for the given arguments, or try to display suggestions for that name
```
```suggestion
        Otherwise display the tag if one was found for the given arguments, or try to display suggestions for that name.
```

Numerlor commented:
I decided to remove the converter for tag names in f5db0b8 as after the removal of adding tags through the bot, it just restricted the input which can safely be ignored by the cog and with the new handling it could refuse tags when parsing a message even if the unparseable part of the message wasn't a part of the tag identifier.

An error caused by that was fixed in f3634f9dbcb6c7cc6952f7d9e40879518c4e6eb1 but I believe that should be everything that could have problems with input that wouldn't pass through the converter before

Bluenix2 commented:
Sending a reminder about [these messages from the Discord server](https://discord.com/channels/267624335836053506/635950537262759947/885576813278011452) so that it is also on record in the PR.

Numerlor commented:
> Sending a reminder about [these messages from the Discord server](https://discord.com/channels/267624335836053506/635950537262759947/885576813278011452) so that it is also on record in the PR.

Cleaned it up a bit in https://github.com/python-discord/bot/pull/1663/files/e07febdbde4815466b161b535f4a5eaf3593f755..3edce6120b48cf83f8a87f34da1007221967232c after I realized the `TagIdentifier.from_string` handles part of the logic and can be used instead. I've also moved all the paginating into the command itself, which makes it a bit busier but I think the interface of the methods that used it makes more sense in that way when it also returns like in your suggestion.

Unfortunately there's one issue with the new handling where I'm not sure how to modify the command help of the group so it's formatted like the get command through the `usage` kwarg
![](https://i.imgur.com/MO4JsIT.png)

Numerlor commented on bot/exts/info/tags.py:
```diff
@@ -172,123 +248,146 @@ async def search_tag_content(self, ctx: Context, *, keywords: str) -> None:
+            if identifier.group == group and tag.accessible_by(user)
+        )
+
+    @tags_group.command(name="get", aliases=("show", "g"), usage="[tag_group] [tag_name]")
```
Do you think `[[tag_group] <tag_name>]"` would make more sense for the usage? to highlight that they're either both optional, or the name is required and the group is optional

Bluenix2 replied:
I don't think people read these carefully enough for it to matter and if they did then that would simply be confusing.

I am open for the discussion but what we have right now seems good enough.

Bluenix2 commented:
> Unfortunately there's one issue with the new handling where I'm not sure how to modify the command help of the group so it's formatted like the get command through the `usage` kwarg
> ![](https://camo.githubusercontent.com/e7aad462d9abeb033ab856ce5c44a1c37314777138e159963d3aaad8a930900d/68747470733a2f2f692e696d6775722e636f6d2f4d4f344a7349542e706e67)

Reading this back, I don't think I follow exactly what issue you're referencing. The brief is the first line in the docstring, so if you can change the docstring to fit enough information in one line then that will work.

Otherwise override the brief (assuming that line is the brief from discord.py).

Numerlor commented:
> Reading this back, I don't think I follow exactly what issue you're referencing. The brief is the first line in the docstring, so if you can change the docstring to fit enough information in one line then that will work.
> 
> Otherwise override the brief (assuming that line is the brief from discord.py).

I was talking about the signature specifically which is a property on the class object, but after reading through the source it became apparent that specifying `usage` does work on groups too and I probably just didn't relaunch the bot or something like that so I through it didn't.
resolved it through a fixup on b56c340

Bluenix2 reviewed (approved):
After countless code-reviews, I have played around with this - very good job here!

Frontmatter, groups and cooldowns all working well! Thanks Numerlor. There are some conflicts to resolve before this can be merged though

onerandomusername commented on tests/bot/exts/backend/test_error_handler.py:
```diff
@@ -359,38 +357,12 @@ async def test_try_get_tag_command_error(self):
     async def test_dont_call_suggestion_tag_sent(self):
         """Should never call command suggestion if tag is already sent."""
-        self.ctx.invoked_with = "foo"
+        self.ctx.message = MagicMock(content="foo")
```
tests/helpers.py has a `MockMessage`, we should use that.

Bluenix2 commented:
Oh I thought this was merged already.. kind of surprised it isn't.

onerandomusername commented:
> Oh I thought this was merged already.. kind of surprised it isn't.

Same here, it looked like it was going to be merged with the conflicts above sorted.

jchristgit reviewed (approved):
:shipit:
